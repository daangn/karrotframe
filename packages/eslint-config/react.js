module.exports = {
  parser: "@typescript-eslint/parser",
  extends: ["airbnb", "@stackflow/eslint-config"],
  rules: {
    "react/no-danger": "off",
    "react/jsx-filename-extension": [
      1,
      { extensions: [".js", ".jsx", ".tsx"] },
    ],
    "react/function-component-definition": "off",
    "react/require-default-props": "off",
    "react/destructuring-assignment": "off",
    "react/prop-types": "off",
    "react/jsx-props-no-spreading": "off",
    "react/react-in-jsx-scope": "off",
  },
};
