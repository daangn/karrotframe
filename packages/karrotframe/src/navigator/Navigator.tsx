import classnames from 'clsx'
import React, { useCallback, useEffect, useRef } from 'react'
import {
  HashRouter,
  matchPath,
  useHistory,
  useLocation,
} from 'react-router-dom'
import { CSSTransition, TransitionGroup } from 'react-transition-group'

import { Card } from './components'
import { NavigatorOptionsProvider, useNavigatorOptions } from './contexts'
import {
  getNavigatorParams,
  NavigatorParamKeys,
  NavigatorTheme,
} from './helpers'
import {
  UniqueIdProvider,
  useHistoryPopEffect,
  useHistoryPushEffect,
  useHistoryReplaceEffect,
  useUniqueId,
} from './hooks'
import styles from './Navigator.scss'
import {
  ScreenInstance,
  StoreProvider,
  useStore,
  useStoreActions,
  useStoreSelector,
} from './store'

declare global {
  interface Window {
    __KARROTFRAME__?: boolean
  }
}

const DEFAULT_CUPERTINO_ANIMATION_DURATION = 350
const DEFAULT_ANDROID_ANIMATION_DURATION = 270

interface NavigatorProps {
  /**
   * Theme (default: `Web`)
   */
  theme?: NavigatorTheme

  /**
   * Transition animation duration
   */
  animationDuration?: number

  /**
   * Remove built-in `<HashRouter />`
   * and use your own custom `<Router />`
   */
  useCustomRouter?: boolean

  /**
   * When close button clicked
   */
  onClose?: () => void

  /**
   * When navigation depth changed
   */
  onDepthChange?: (depth: number) => void
}
const Navigator: React.FC<NavigatorProps> = (props) => {
  let h = (
    <NavigatorOptionsProvider
      value={{
        theme: props.theme ?? 'Android',
        animationDuration:
          props.animationDuration ??
          (() => {
            switch (props.theme ?? 'Android') {
              case 'Cupertino':
                return DEFAULT_CUPERTINO_ANIMATION_DURATION
              case 'Android':
                return DEFAULT_ANDROID_ANIMATION_DURATION
            }
          })(),
      }}
    >
      <UniqueIdProvider>
        <StoreProvider>
          <NavigatorScreens
            theme={props.theme ?? 'Android'}
            onClose={props.onClose}
            onDepthChange={props.onDepthChange}
          >
            {props.children}
          </NavigatorScreens>
        </StoreProvider>
      </UniqueIdProvider>
    </NavigatorOptionsProvider>
  )

  if (!props.useCustomRouter) {
    h = <HashRouter>{h}</HashRouter>
  }

  return h
}

interface NavigatorScreensProps {
  theme: NavigatorTheme
  onClose?: () => void
  onDepthChange?: (depth: number) => void
}
const NavigatorScreens: React.FC<NavigatorScreensProps> = (props) => {
  const location = useLocation()
  const history = useHistory()
  const { uid } = useUniqueId()

  const store = useStore()
  const { screenInstances, screenInstancePtr } = useStoreSelector((state) => ({
    screenInstances: state.screenInstances,
    screenInstancePtr: state.screenInstancePtr,
  }))
  const {
    increaseScreenInstancePtr,
    insertScreenInstance,
    mapScreenInstance,
    setScreenInstancePtr,
  } = useStoreActions()

  const pushScreen = useCallback(
    ({
      screenId,
      screenInstanceId,
      present,
      as,
    }: {
      screenId: string
      screenInstanceId: string
      present: boolean
      as: string
    }) => {
      const { screenInstances, screenInstancePtr } = store.getState()

      const nextPtr = screenInstances.findIndex(
        (screenInstance) => screenInstance.id === screenInstanceId
      )

      if (nextPtr === -1) {
        insertScreenInstance({
          ptr: screenInstancePtr,
          screenInstance: {
            id: screenInstanceId,
            screenId,
            present,
            as,
          },
        })
        increaseScreenInstancePtr()
      } else {
        setScreenInstancePtr({ ptr: nextPtr })
      }
    },
    []
  )

  const replaceScreen = useCallback(
    ({
      screenId,
      screenInstanceId,
      as,
      present,
    }: {
      screenId: string
      screenInstanceId: string
      as: string
      present: boolean
    }) => {
      const { screenInstancePtr } = store.getState()

      insertScreenInstance({
        ptr: screenInstancePtr - 1,
        screenInstance: {
          id: screenInstanceId,
          screenId,
          present,
          as,
        },
      })
    },
    []
  )

  const popScreen = useCallback(
    ({
      depth,
      targetScreenInstanceId,
    }: {
      depth: number
      targetScreenInstanceId?: string
    }) => {
      const { screenInstancePromises, screenInstancePtr } = store.getState()
      if (targetScreenInstanceId) {
        const promise = screenInstancePromises[targetScreenInstanceId]

        if (promise && !promise.popped) {
          promise.resolve(null)
        }
      }
      setScreenInstancePtr({
        ptr: screenInstancePtr - depth,
      })
    },
    []
  )

  useEffect(() => {
    if (typeof window === 'undefined') {
      return
    }
    if (window.__KARROTFRAME__) {
      throw new Error('한 개의 앱에는 한 개의 Navigator만 허용됩니다')
    }

    window.__KARROTFRAME__ = true

    const searchParams = new URLSearchParams(location.search)
    searchParams.set(NavigatorParamKeys.screenInstanceId, uid())

    history.replace(`${location.pathname}?${searchParams.toString()}`)

    return () => {
      window.__KARROTFRAME__ = false
    }
  }, [])

  useEffect(() => {
    if (!location.search) {
      return
    }

    const { screens, screenInstances } = store.getState()

    if (screenInstances.length > 0) {
      return
    }

    const searchParams = new URLSearchParams(location.search)
    const { screenInstanceId } = getNavigatorParams(searchParams)

    const matchScreen = Object.values(screens).find(
      (screen) =>
        screen &&
        matchPath(location.pathname, { exact: true, path: screen.path })
    )

    if (screenInstanceId && matchScreen) {
      pushScreen({
        screenId: matchScreen.id,
        screenInstanceId,
        present: false,
        as: location.pathname,
      })
    }
  }, [location.search])

  useEffect(() => {
    return store.listen((prevState, nextState) => {
      if (
        nextState.screenInstancePtr > -1 &&
        prevState.screenInstancePtr !== nextState.screenInstancePtr
      ) {
        props.onDepthChange?.(nextState.screenInstancePtr)
      }
    })
  }, [props.onDepthChange])

  useHistoryPushEffect(
    (location) => {
      const { screens, screenInstancePtr } = store.getState()

      const searchParams = new URLSearchParams(location.search)
      const { screenInstanceId, present } = getNavigatorParams(searchParams)

      const matchScreen = Object.values(screens).find(
        (screen) =>
          !!screen &&
          matchPath(location.pathname, { exact: true, path: screen.path })
      )

      if (screenInstanceId && matchScreen) {
        pushScreen({
          screenId: matchScreen.id,
          screenInstanceId,
          present,
          as: location.pathname,
        })
      } else {
        mapScreenInstance({
          ptr: screenInstancePtr,
          mapper: (screenInstance) => ({
            ...screenInstance,
            nestedRouteCount: screenInstance.nestedRouteCount + 1,
          }),
        })
      }
    },
    [pushScreen]
  )

  useHistoryReplaceEffect(
    (location) => {
      const { screens } = store.getState()

      const searchParams = new URLSearchParams(location.search)
      const { screenInstanceId, present } = getNavigatorParams(searchParams)

      const matchScreen = Object.values(screens).find(
        (screen) =>
          screen &&
          matchPath(location.pathname, { exact: true, path: screen.path })
      )

      if (screenInstanceId && matchScreen) {
        replaceScreen({
          screenId: matchScreen.id,
          screenInstanceId,
          present,
          as: location.pathname,
        })
      }
    },
    [replaceScreen]
  )

  useHistoryPopEffect(
    {
      backward(location) {
        const { screens, screenInstances, screenInstancePtr } = store.getState()

        const matchScreen = Object.values(screens).find(
          (screen) =>
            screen &&
            matchPath(location.pathname, { exact: true, path: screen.path })
        )

        const searchParams = new URLSearchParams(location.search)
        const { screenInstanceId } = getNavigatorParams(searchParams)

        if (screenInstanceId && matchScreen) {
          const nextPtr = screenInstances.findIndex(
            (screenInstance) => screenInstance.id === screenInstanceId
          )

          mapScreenInstance({
            ptr: screenInstancePtr,
            mapper: (screenInstance) => ({
              ...screenInstance,
              nestedRouteCount: 0,
            }),
          })
          popScreen({
            depth: screenInstancePtr - nextPtr,
            targetScreenInstanceId: screenInstanceId,
          })
        } else if (screenInstances[screenInstancePtr]?.nestedRouteCount === 0) {
          popScreen({
            depth: 1,
          })
        } else {
          mapScreenInstance({
            ptr: screenInstancePtr,
            mapper: (screenInstance) => ({
              ...screenInstance,
              nestedRouteCount: screenInstance.nestedRouteCount - 1,
            }),
          })
        }
      },
      forward(location) {
        const { screens, screenInstancePtr } = store.getState()
        const searchParams = new URLSearchParams(location.search)
        const { screenInstanceId, present } = getNavigatorParams(searchParams)

        const matchScreen = Object.values(screens).find(
          (screen) =>
            screen &&
            matchPath(location.pathname, { exact: true, path: screen.path })
        )

        if (screenInstanceId && matchScreen) {
          pushScreen({
            screenId: matchScreen.id,
            screenInstanceId,
            present,
            as: location.pathname,
          })
        } else {
          mapScreenInstance({
            ptr: screenInstancePtr,
            mapper(screenInstance: ScreenInstance): ScreenInstance {
              return {
                ...screenInstance,
                nestedRouteCount: screenInstance.nestedRouteCount + 1,
              }
            },
          })
        }
      },
    },
    [popScreen, pushScreen]
  )

  return (
    <div
      className={classnames(styles.navigatorRoot, {
        'kf-android': props.theme === 'Android',
        'kf-cupertino': props.theme === 'Cupertino',
      })}
    >
      {props.children}
      <TransitionGroup component={null}>
        {screenInstances.map((screenInstance, index) => (
          <Transition
            key={index}
            screenInstance={screenInstance}
            screenInstanceIndex={index}
            isRoot={index === 0}
            isTop={index === screenInstancePtr}
            onClose={props.onClose}
          />
        ))}
      </TransitionGroup>
    </div>
  )
}

interface TransitionProps {
  screenInstance: ScreenInstance
  screenInstanceIndex: number
  isRoot: boolean
  isTop: boolean
  onClose?: () => void
}
const Transition: React.FC<TransitionProps> = (props) => {
  const navigatorOptions = useNavigatorOptions()
  const nodeRef = useRef<HTMLDivElement>(null)

  const { screens, screenInstancePtr, screenInstances } = useStoreSelector(
    (state) => ({
      screens: state.screens,
      screenInstancePtr: state.screenInstancePtr,
      screenInstances: state.screenInstances,
    })
  )

  const screen = screens[props.screenInstance.screenId]

  if (!screen) {
    return null
  }

  return (
    <CSSTransition
      key={props.screenInstance.id}
      nodeRef={nodeRef}
      timeout={navigatorOptions.animationDuration}
      in={props.screenInstanceIndex <= screenInstancePtr}
      unmountOnExit
    >
      <Card
        nodeRef={nodeRef}
        screenPath={screen.path}
        screenInstanceId={props.screenInstance.id}
        isRoot={props.screenInstanceIndex === 0}
        isTop={
          props.screenInstanceIndex >= screenInstancePtr ||
          (navigatorOptions.theme === 'Cupertino' &&
            screenInstances.length > props.screenInstanceIndex + 1 &&
            screenInstances[props.screenInstanceIndex + 1].present)
        }
        isPresent={props.screenInstance.present}
        onClose={props.onClose}
      >
        <screen.Component
          as={props.screenInstance.as}
          screenInstanceId={props.screenInstance.id}
          isTop={props.isTop}
          isRoot={props.isRoot}
        />
      </Card>
    </CSSTransition>
  )
}

export default Navigator
