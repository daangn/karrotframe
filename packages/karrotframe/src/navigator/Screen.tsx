import React, { useEffect } from 'react'

import {
  ScreenInstanceInfoProvider,
  ScreenInstanceOptionsProvider,
} from './contexts'
import { ScreenComponentProps } from './ScreenComponentProps'
import { NavbarOptions, useStoreActions } from './store'

interface Props {
  /**
   * URL path
   */
  path: string

  /**
   * Component
   */
  component?: React.ComponentType<ScreenComponentProps>
}
const Screen: React.FC<Props> = (props) => {
  const { component: Component } = props
  const { addScreen, addScreenInstanceOption, removeScreen } = useStoreActions()

  useEffect(() => {
    if (!props.children && !Component) {
      console.warn('component props, children 중 하나는 반드시 필요합니다')
      return
    }

    const screenId = props.path

    addScreen({
      screen: {
        id: screenId,
        path: props.path,
        Component({ screenInstanceId, isTop, isRoot, as }) {
          const setNavbar = (navbar: NavbarOptions) => {
            addScreenInstanceOption({
              screenInstanceId,
              screenInstanceOption: {
                navbar,
              },
            })
          }

          return (
            <ScreenInstanceInfoProvider
              value={{
                screenInstanceId,
                as,
                path: props.path,
              }}
            >
              <ScreenInstanceOptionsProvider
                value={{
                  setNavbar,
                }}
              >
                {Component ? (
                  <Component isTop={isTop} isRoot={isRoot} />
                ) : (
                  props.children
                )}
              </ScreenInstanceOptionsProvider>
            </ScreenInstanceInfoProvider>
          )
        },
      },
    })

    return () => {
      removeScreen({
        screenId,
      })
    }
  }, [props.path, Component])

  return null
}

export default Screen
