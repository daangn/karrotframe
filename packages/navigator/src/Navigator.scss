.navigatorRoot {
  width: 100%;
  height: 100%;
  overflow: hidden;
  position: relative;
  user-select: none;
}
