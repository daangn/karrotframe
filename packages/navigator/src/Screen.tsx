import React, { useCallback, useEffect, useMemo } from 'react'

import {
  ScreenInstanceProvider,
  ScreenInstanceSetNavbarProvider,
} from './contexts'
import { INavbarOptions, useStoreActions } from './store'

interface IScreenProps {
  /**
   * URL path
   */
  path: string

  /**
   * Component
   */
  component?: React.ComponentType
}
const Screen: React.FC<IScreenProps> = (props) => {
  const { component: Component } = props
  const { registerScreen, addScreenInstanceOption } = useStoreActions()

  useEffect(() => {
    if (!props.children && !Component) {
      console.warn('Either component props or children is required')
      return
    }

    const screenId = props.path

    const unregisterScreen = registerScreen({
      screen: {
        id: screenId,
        path: props.path,
        Component({ screenInstanceId, as, isTop, isRoot }) {
          const setNavbar = useCallback(
            (navbar: INavbarOptions) => {
              addScreenInstanceOption({
                screenInstanceId,
                screenInstanceOption: {
                  navbar,
                },
              })
            },
            [addScreenInstanceOption, screenInstanceId]
          )

          const screenInstanceContext = useMemo(
            () => ({
              screenInstanceId,
              as,
              isTop,
              isRoot,
              path: props.path,
            }),
            [screenInstanceId, as, isTop, isRoot, props.path]
          )

          return (
            <ScreenInstanceProvider value={screenInstanceContext}>
              <ScreenInstanceSetNavbarProvider value={setNavbar}>
                {Component ? <Component /> : props.children}
              </ScreenInstanceSetNavbarProvider>
            </ScreenInstanceProvider>
          )
        },
      },
    })

    return () => {
      unregisterScreen()
    }
  }, [props.path, Component])

  return null
}

export default Screen
