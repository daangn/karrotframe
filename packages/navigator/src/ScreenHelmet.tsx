import React, { useEffect } from 'react'

import {
  makeScreenHelmetDefaultProps,
  useScreenHelmet,
} from './components/Stack.ContextScreenHelmet'

export interface IScreenHelmetProps {
  /**
   * title
   */
  title?: React.ReactNode

  /**
   * Append elements in left side
   * (It'll be displayed in right side of back button)
   */
  appendLeft?: React.ReactNode

  /**
   * Append elements in right side
   * (It'll be displayed in left side of close button when `closeButtonLocation` is `right`)
   */
  appendRight?: React.ReactNode

  /**
   * Location of close button (default: `left`)
   */
  closeButtonLocation?: 'left' | 'right'

  /**
   * Replace back button
   */
  customBackButton?: React.ReactNode

  /**
   * Replace close button
   */
  customCloseButton?: React.ReactNode

  /**
   * Remove border
   */
  noBorder?: boolean

  /**
   * Disable scroll to top feature (iOS Only)
   */
  disableScrollToTop?: boolean

  /**
   * When top part clicked (iOS Only)
   */
  onTopClick?: () => void

  /**
   * Set visibility for NavBar (default: `true`)
   */
  visible?: boolean

  /**
   * block event when users try to swipe back
   */
  preventSwipeBack?: boolean
  /**
   * hide button from leftside of navbar in ScreenHelmet
   */
  hideLeftButton?: boolean
}
const ScreenHelmet: React.FC<IScreenHelmetProps> = (props) => {
  const {
    setScreenHelmetProps,
    setScreenHelmetVisible,
    resetScreenHelmetProps,
  } = useScreenHelmet()

  useEffect(() => {
    setScreenHelmetProps({
      ...makeScreenHelmetDefaultProps(),
      ...Object.entries(props).reduce(
        (a, [k, v]) => (v === undefined || v === null ? a : ((a[k] = v), a)),
        {}
      ),
    })
    setScreenHelmetVisible(props.visible ?? true)
  }, [props])

  useEffect(() => resetScreenHelmetProps, [resetScreenHelmetProps])

  return null
}

export default ScreenHelmet
