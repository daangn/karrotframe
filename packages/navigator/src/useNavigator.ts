import {useCallback, useMemo } from 'react'
import { useHistory, useLocation } from 'react-router-dom'

import { useScreenInstance } from './components/Stack.ContextScreenInstance'
import { useScreenInstances } from './globalState'
import {
  makeNavigatorSearchParams,
  nextTick,
  parseNavigatorSearchParams,
} from './helpers'
import { useIncrementalId } from './hooks'

enum PluginTarget {
    screenInstance
}

export interface PluginType {
    target:PluginTarget;
    isDataRequired: boolean;
    handlers: {[funName: string]: () => void;};
    listeners: {
        beforePush?: (to: string) => void;
        onPushed?: (to: string) => void;
        beforePop?: (from: string) => void;
        onPopped?: (from: string, data: any, options: any) => void;
    }
}

export function useNavigator() {
  const history = useHistory()
  const location = useLocation()
  const screenInfo = useScreenInstance()
  const makeId = useIncrementalId()

  const {
    screenInstances,
    screenInstancePtr,
    screenInstancePromiseMap,
    addScreenInstancePromise,
    screenPlugins
  } = useScreenInstances()

  const navigatorSearchParams = parseNavigatorSearchParams(location.search)
  const { present, screenInstanceId } = navigatorSearchParams.toObject()

  const currentScreenInstance = useMemo(() => screenInstances[screenInstancePtr], [screenInstances, screenInstancePtr]);
  const dataPlugin = useMemo(() =>  screenPlugins.find(plugin => plugin.isDataRequired), [screenPlugins])

  const beforePush = useCallback((to: string) => {
      screenPlugins.forEach(plugin => {
          plugin?.listeners?.beforePush?.(to);
      })
  }, [screenPlugins])

  const onPushed = useCallback((to) => {
      screenPlugins.forEach(plugin => {
              plugin?.listeners.onPushed?.(to);
      })
  }, [screenPlugins])

  const push = useCallback(
    <T = object>(
      to: string,
      options?: {
        /**
         * Bottom to top animation (iOS only)
         */
        present?: boolean
      },
    ): Promise<T | null> =>
      new Promise((resolve) => {
        beforePush(to);
        const { pathname, searchParams } = new URL(to, /* dummy */ 'file://')

        const navigatorSearchParams = makeNavigatorSearchParams(searchParams, {
          screenInstanceId: makeId(),
          present: options?.present,
        })

        addScreenInstancePromise({
          screenInstanceId: screenInfo.screenInstanceId,
          screenInstancePromise: {
            resolve,
          },
        })
        onPushed(to);
        history.push(`${pathname}?${navigatorSearchParams.toString()}`)
      }),
    [screenInfo, history]
  )

  const replace = useCallback(
    (
      to: string,
      options?: {
        /**
         * Animate when replaced
         */
        animate?: boolean
      }
    ) => {
      const { pathname, searchParams } = new URL(to, /* dummy */ 'file://')

      const navigatorSearchParams = makeNavigatorSearchParams(searchParams, {
        screenInstanceId: options?.animate ? makeId() : screenInstanceId,
        present,
      })

      nextTick(() => {
        history.replace(`${pathname}?${navigatorSearchParams.toString()}`)
      })
    },
    [history, screenInstanceId, present]
  )

    const beforePop = useCallback(() => {
        screenPlugins.forEach(plugin => {
            plugin?.listeners.beforePop?.(currentScreenInstance.as);
        })
    }, [])

    const onPopped = useCallback(() => {
        screenPlugins.forEach(plugin => {
            // TODO: prepare data and options, unifying onPoppedWithData
            plugin?.listeners.onPopped?.(currentScreenInstance.as, null, null);
        })

    }, [])

    const onPoppedWithData = useCallback((data: any) => {
        dataPlugin?.listeners.onPopped?.(currentScreenInstance.as, data, {});
    }, [])

  const pop = useCallback(
    (depth = 1) => {
        beforePop();
      const targetScreenInstance = screenInstances[screenInstancePtr - depth]

      const backwardCount = screenInstances
        .filter(
          (_, idx) =>
            idx > screenInstancePtr - depth && idx <= screenInstancePtr
        )
        .map((screenInstance) => screenInstance.nestedRouteCount)
        .reduce((acc, current) => acc + current + 1, 0)

      const targetPromise =
        targetScreenInstance &&
        screenInstancePromiseMap[targetScreenInstance.id]
      let _data: any = null

      const dispose = history.listen(() => {
        dispose()

        if (targetScreenInstance && targetPromise?.resolve) {
          targetPromise.resolve(_data ?? null)
        }
      })

      /**
       * Send data to `await push()`
       */
      function send<T = object>(
        /**
         * Payload
         */
        data: T,
      ) {
        _data = data
        // FIXME: 'onPoppedWithData' and 'onPopped' should be unified later
        onPoppedWithData(data);
      }
      onPopped();
      nextTick(() => {
        history.go(-backwardCount)
      })

      return {
        send,
      }
    },
    [screenInstances, screenInstancePtr, screenInstancePromiseMap, history]
  )

  return useMemo(
    () => ({
      push,
      replace,
      pop,
      ...screenPlugins.reduce((acc, curr) => {
            if(curr && curr.handlers) {
                acc = {
                    ...acc,
                    ...Object.keys(curr.handlers).reduce((_acc, _curr) => {
                        _acc = {
                            ..._acc,
                            [_curr]: curr.handlers[_curr]
                        }
                        return _acc;
                    }, {}),
                }
            }
            return acc;
        }, {})
    }),
    [push, replace, pop]
  )
}
