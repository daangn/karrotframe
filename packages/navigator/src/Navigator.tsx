import React from 'react'
import { HashRouter } from 'react-router-dom'
import { TransitionGroup } from 'react-transition-group'

import { assignInlineVars } from '@vanilla-extract/dynamic'

import Stack from './components/Stack'
import { ProviderScreenInstances, ProviderScreens } from './globalState'
import { ProviderIncrementalId } from './hooks'
import * as css from './Navigator.css'
import { INavigatorTheme } from './types'
import type { NavigatorPluginType } from '@karrotframe/plugins'
import wrapProvider from './helpers/wrapProvider'
import { ProviderPlugins } from './globalState/Plugins'

declare global {
  interface Window {
    __KARROTFRAME__?: boolean
  }
}

const DEFAULT_CUPERTINO_ANIMATION_DURATION = 350
const DEFAULT_ANDROID_ANIMATION_DURATION = 270

interface INavigatorProps {
  /**
   * Theme (default: `Android`)
   */
  theme?: INavigatorTheme

  /**
   * Transition animation duration
   */
  animationDuration?: number

  /**
   * Remove built-in `<HashRouter />`
   * and use your own custom `<Router />`
   */
  useCustomRouter?: boolean

  /**
   * Class name appended to root div element
   */
  className?: string

  /**
   * `aria-label=` property assigned to back button
   */
  backButtonAriaLabel?: string

  /**
   * `aria-label=` property assigned to close button
   */
  closeButtonAriaLabel?: string

  /**
   * When close button clicked
   */
  onClose?: () => void

  /**
   * When navigation depth changed
   */
  onDepthChange?: (depth: number) => void

  /**
   * plugins
   */
  plugins?: NavigatorPluginType[]
}
const Navigator: React.FC<INavigatorProps> = ({
  theme = 'Android',
  animationDuration = theme === 'Android'
    ? DEFAULT_ANDROID_ANIMATION_DURATION
    : DEFAULT_CUPERTINO_ANIMATION_DURATION,
  useCustomRouter,
  className,
  backButtonAriaLabel = 'Go back',
  closeButtonAriaLabel = 'Close',
  onClose,
  onDepthChange,
  plugins = [],
  children,
}) => {
  let h = (
    <ProviderIncrementalId>
      {wrapProvider(
        plugins.map((plugin) => plugin.provider).filter(Boolean),
        () => (
          <ProviderPlugins plugins={plugins}>
            <ProviderScreens>
              <ProviderScreenInstances>
                <div
                  className={[
                    css.root({ theme }),
                    ...(className ? [className] : []),
                  ].join(' ')}
                  style={assignInlineVars({
                    [css.vars.animationDuration]: animationDuration + 'ms',
                  })}
                >
                  <TransitionGroup component={null}>
                    <Stack
                      animationDuration={animationDuration}
                      theme={theme}
                      onClose={onClose}
                      backButtonAriaLabel={backButtonAriaLabel}
                      closeButtonAriaLabel={closeButtonAriaLabel}
                      onDepthChange={onDepthChange}
                    >
                      {children}
                    </Stack>
                  </TransitionGroup>
                </div>
              </ProviderScreenInstances>
            </ProviderScreens>
          </ProviderPlugins>
        )
      )}
    </ProviderIncrementalId>
  )

  if (!useCustomRouter) {
    h = <HashRouter>{h}</HashRouter>
  }

  return h
}

export default Navigator
