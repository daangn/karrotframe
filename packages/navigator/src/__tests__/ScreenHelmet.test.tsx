import React from 'react'
import type { FC, ReactElement } from 'react'
import { render, fireEvent, waitFor, screen } from '@testing-library/react'
import type { RenderResult } from '@testing-library/react'

import ScreenHelmet from '../ScreenHelmet'
import Navigator from '../Navigator'
import Screen from '../Screen'
import { useNavigator } from '../useNavigator'

describe('ScreenHelmet - visible: ', () => {
  const renderScreenHelmet = ({
    visible,
  }: {
    visible: boolean
  }): RenderResult => {
    const Example: FC = (): ReactElement => {
      return (
        <div>
          <ScreenHelmet visible={visible} />
          <span>example</span>
        </div>
      )
    }

    return render(
      <Navigator>
        <Screen path="/" component={Example} />
      </Navigator>
    )
  }

  it('visible: false 이면 navbar 가 나타나지 않는다', () => {
    const { queryByTestId } = renderScreenHelmet({ visible: false })
    const navBar = queryByTestId('navbar')
    expect(navBar).not.toBeInTheDocument()
  })

  it('visible: true 이면 navbar 가 나타난다', () => {
    const { getByTestId } = renderScreenHelmet({ visible: true })
    const navBar = getByTestId('navbar')
    expect(navBar).toBeVisible()
  })
})

describe('ScreenHelmet - preventSwipeBack:  ', () => {
  const renderScreenHelmet = ({
    preventSwipeBack,
  }: {
    preventSwipeBack: boolean
  }): RenderResult => {
    const Example: FC = (): ReactElement => {
      const { push } = useNavigator()

      return (
        <div>
          <ScreenHelmet preventSwipeBack={preventSwipeBack} />
          <span>example</span>
          <button
            onClick={() => {
              push('/another')
            }}
          >
            move
          </button>
        </div>
      )
    }
    const Another: FC = (): ReactElement => {
      const { push } = useNavigator()

      return (
        <div>
          <ScreenHelmet preventSwipeBack={preventSwipeBack} />
          <span>another</span>
          <button
            onClick={() => {
              push('/')
            }}
          >
            teardown
          </button>
        </div>
      )
    }

    return render(
      <Navigator theme="Cupertino">
        <Screen path="/" component={Example} />
        <Screen path="/another" component={Another} />
      </Navigator>
    )
  }

  afterEach(async () => {
    try {
      await waitFor(() => {
        const teardownButton = screen.queryByText(/teardown/i)
        if (teardownButton) {
          fireEvent.click(teardownButton)
        }
      })
    } catch (e) {
      console.error(e)
    }
  })

  it('preventSwipeBack: true 이면 edge element 를 생성하지 않는다 ', async () => {
    // when
    const { getByText, queryByTestId } = renderScreenHelmet({
      preventSwipeBack: true,
    })
    const moveButton = getByText(/move/i)
    fireEvent.click(moveButton)

    // then
    await waitFor(() => {
      const edgeElement = queryByTestId('edge-element')
      expect(edgeElement).not.toBeInTheDocument()
    })
  })

  it('preventSwipeBack: false 이면 edge element 를 생성한다. ', async () => {
    // when
    const { getByText, findByTestId } = renderScreenHelmet({
      preventSwipeBack: false,
    })
    const moveButton = getByText(/move/i)
    fireEvent.click(moveButton)

    // then
    const edgeElement = await findByTestId('edge-element')
    expect(edgeElement).toBeInTheDocument()
  })
})
