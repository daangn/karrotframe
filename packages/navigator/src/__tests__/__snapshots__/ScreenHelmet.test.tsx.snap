// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScreenHelmet - visible:  visible: false 이면 navbar 가 나타나지 않는다 1`] = `
<DocumentFragment>
  <div
    class="Navigator_root__hije1d2 Navigator_root_Android__hije1d3 Navigator_Android__hije1d0"
    style="--kf_navigator_animationDuration: 270ms;"
  >
    <div
      class="Card_container__1ok9rul0 Card_container_enterActive__1ok9rul1"
    >
      <div
        class="Card_mainOffset__1ok9rul9"
      >
        <div
          class="Card_main__1ok9rulc Card_main_true__1ok9rulh Card_main_true__1ok9rulj"
          style="--kf_navigator_navbar-animationDuration: 0s;"
        >
          <div
            class="Card_frameOffset__1ok9rull"
          >
            <div
              class="Card_frame__1ok9rulp"
            >
              <div>
                <span>
                  example
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ScreenHelmet - visible:  visible: true 이면 navbar 가 나타난다 1`] = `
<DocumentFragment>
  <div
    class="Navigator_root__hije1d2 Navigator_root_Android__hije1d3 Navigator_Android__hije1d0"
    style="--kf_navigator_animationDuration: 270ms;"
  >
    <div
      class="Card_container__1ok9rul0 Card_container_enterActive__1ok9rul1"
    >
      <div
        class="Card_mainOffset__1ok9rul9"
      >
        <div
          class="Card_main__1ok9rulc Card_main_true__1ok9rulh Card_main_true__1ok9ruli Card_main_true__1ok9rulj"
          style="--kf_navigator_navbar-animationDuration: 0s;"
        >
          <div
            class="Navbar_container__x613ng0"
            data-testid="navbar"
            style="--kf_navigator_navbar-center-mainWidth: undefinedpx;"
          >
            <div
              class="Navbar_main__x613ng2"
            >
              <div
                class="Navbar_flex__x613ng4"
              >
                <div
                  class="Navbar_left__x613ng5"
                >
                  <a
                    aria-label="Close"
                    class="Navbar_closeButton__x613ng9"
                    role="text"
                  >
                    <svg
                      class="Navbar_svgIcon__x613ngj"
                      fill="none"
                      height="24"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M3.72633 3.72633C4.0281 3.42456 4.51736 3.42456 4.81913 3.72633L12 10.9072L19.1809 3.72633C19.4826 3.42456 19.9719 3.42456 20.2737 3.72633C20.5754 4.0281 20.5754 4.51736 20.2737 4.81913L13.0928 12L20.2737 19.1809C20.5754 19.4826 20.5754 19.9719 20.2737 20.2737C19.9719 20.5754 19.4826 20.5754 19.1809 20.2737L12 13.0928L4.81913 20.2737C4.51736 20.5754 4.0281 20.5754 3.72633 20.2737C3.42456 19.9719 3.42456 19.4826 3.72633 19.1809L10.9072 12L3.72633 4.81913C3.42456 4.51736 3.42456 4.0281 3.72633 3.72633Z"
                        fill="currentColor"
                      />
                    </svg>
                  </a>
                </div>
                <div
                  class="Navbar_center__x613nga Navbar_center_true__x613ngb"
                >
                  <div
                    class="Navbar_centerMain__x613ngc Navbar_centerMain_true__x613ngd Navbar_centerMain_true__x613nge"
                  />
                  <div
                    class="Navbar_centerMainEdge__x613ngh"
                  />
                </div>
                <div
                  class="Navbar_right__x613ng7 Navbar_right_true__x613ng8"
                />
              </div>
            </div>
          </div>
          <div
            class="Card_frameOffset__1ok9rull"
          >
            <div
              class="Card_frame__1ok9rulp"
            >
              <div>
                <span>
                  example
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
